jsonrpc/ipc.go

```diff
@@ -0,0 +1,221 @@
+}
+
+// NewIpc creates a new IPC handler instance with the provided RPC server, endpoint and logger.
+func NewIpc(rpc *Server, endpoint string, log utils.SimpleLogger) *Ipc {
+	return &Ipc{
+		endpoint:       endpoint,
+		rpc:            rpc,
+		log:            log,
+		connParams:     DefaultIpcConnParams(),
+		conns:          make(map[net.Conn]struct{}),
+		events:         &SelectiveListener{},
+		socketCallback: func() {},
+	}
+}
+
+// WithSocketCallback sets the callback for socket creation.
+func (i *Ipc) WithSocketCallback(f func()) *Ipc {
+	i.socketCallback = f
+	return i
+}
+
+// WithConnParams sanity checks and applies the provided params.
+func (i *Ipc) WithConnParams(p *IpcConnParams) *Ipc {
+	i.connParams = *p
+	return i
+}
+
+// WithListener registers a NewRequestListener
+func (i *Ipc) WithListener(listener NewRequestListener) *Ipc {
+	i.events = listener
+	return i
+}
+
+// Run launches the IPC handler and handles any potential errors.
+// It is the caller's responsibility to cancel the provided context when they wish to
+// gracefully shut down the IPC handler.
+func (i *Ipc) Run(ctx context.Context) error {
+	if err := i.initialiseSocket(); err != nil {
+		return err
+	}
+
+	var (
+		errCh = make(chan error, 1)
+		wg    conc.WaitGroup
+	)
+	defer wg.Wait()
+	wg.Go(func() { errCh <- i.run(ctx) })
+
+	select {
+	case <-ctx.Done():
+	case err := <-errCh:
+		if !isSocketError(err) {
+			i.log.Warnw("Listener failure", "err", err)
+		}
+	}
+	return i.stop()
+}
+
+// initialiseSocket initialises the socket and executes the socket callback upon successful creation.
+func (i *Ipc) initialiseSocket() error {
+	listener, err := createListener(i.endpoint)
+	if err != nil {
+		return err
+	}
+	i.listener = listener
+	i.socketCallback()
+	return nil
+}
+
+// run continuously accepts and serves incoming connections.
+func (i *Ipc) run(ctx context.Context) error {
+	for {
+		conn, err := i.listener.Accept()
+		if netutil.IsTemporaryError(err) {
+			i.log.Errorw("Failed to accept connection", "err", err)
+			continue
+		} else if err != nil {
+			return err
+		}
+		i.connWg.Add(1)
+		go i.serveConn(ctx, newIpcConn(conn, i.connParams))
+	}
+}
+
+// setupConn adds a connection and checks for handler closing.
+func (i *Ipc) setupConn(conn net.Conn) bool {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	i.conns[conn] = struct{}{}
+	return !i.closing
+}
+
+// cleanupConn calls cleanupConnNoLock with explicit lock
+func (i *Ipc) cleanupConn(conn net.Conn) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	i.cleanupConnNoLock(conn)
+}
+
+// cleanupConnNoLock frees resources
+func (i *Ipc) cleanupConnNoLock(conn net.Conn) {
+	_, ok := i.conns[conn]
+	delete(i.conns, conn)
+	if ok {
+		conn.Close()
+		i.connWg.Done()
+	}
+}
+
+// serveConn handles incoming connection.
+func (i *Ipc) serveConn(ctx context.Context, conn net.Conn) {
+	defer i.cleanupConn(conn)
+	if !i.setupConn(conn) {
+		return
+	}
```

Comment:
When will this execute?

Reply:
When the handler is stopping and there are still pending connections yet to be scheduled.

Reply:
To be more precise, There can be some connections yet to be scheduled from `go i.serveConn(ctx, newIpcConn(conn, i.connParams))` since closing socket doesn't close the connections.
```go
**// Close stops listening on the Unix address. Already accepted
// connections are not closed.
func (l *UnixListener) Close() error {**
```

Reply:
There is no guarantee that the `HandleReadWriter` will error. Already accepted connection is perfectly fine when `net.Listener` is closed which means that reads/writes are going to work.

Reply:
I see, thanks for clarifying. 

Even still, `closing` may become `true` right after this condition fails, so we may accept the connection anyways. If that's correct, can we remove the `closing` field entirely?

Reply:
If it becomes true afterwards then stop will close that connection. Stop takes over lock and closes every connection, also marks that field as true. After the lock is released every new connection is ensured to see true. That also means that every connection before stop will be closed by that or either when connection errors (the cleanup thats deferred also takes the same lock and closes if its still there)

Reply:
> If it becomes true afterwards then stop will close that connection.

Right, this is why I think we are better off removing the `i.closing` condition anyway. You mention that stopping the connection while we are reading from it causes an error. Will that be caught at the `if isSocketErr(err)...` line after the `for` loop? If so, I don't see any downsides to removing the `i.closing` field.

Reply:
Yes, that will be caught by that. What I am trying to say is that there can be such a case:
1. It closes the listener which rejects all of the new connections
2. Some goroutines are yet to be scheduled from go runtime (go serveConn can be executed some time in the future)

Reply:
The i.stopping is actually stopped signal meaning that the process of closing and removing all connections has already ended. In that case that connection will never be closed untill it errors out but it will only error out when either  juno encounters error internally on some rpc method or when client disconnects.

Reply:
Ahhh I see now. That's a tricky edge case.

I think using a child context as proposed in https://github.com/NethermindEth/juno/pull/1362#discussion_r1425831083 can help us here.
```go
	i.mu.Lock()
	i.conns[conn] = struct{}{}
	i.mu.Unlock()
	if ctx.Err() != nil {
		return
	}
```
Right now, there are two ways that `i.stop` is called: `ctx` is canceled or `ch` is closed. If we have a single child context, we simplify the invariant to: `i.stop` is called when `childCtx` is canceled. Since the child context is passed to `serveConn`, we can check it instead of `i.closing`.

Does this sound alright?